<a href="https://colab.research.google.com/github/Abir-Reza/MachineLearning_with_Tensorflow/blob/master/fashionmnist_with_Convolution.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Training Without Convolution

In [11]:
import tensorflow as tf
mnist = tf.keras.datasets.fashion_mnist
(training_images, training_labels), (test_images, test_labels) = mnist.load_data()

training_images=training_images / 255.0
test_images=test_images / 255.0

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(128, activation=tf.nn.relu),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
model.fit(training_images, training_labels, epochs=5)

test_loss = model.evaluate(test_images, test_labels)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.6233 - accuracy: 0.7875
Epoch 2/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3868 - accuracy: 0.8616
Epoch 3/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3390 - accuracy: 0.8747
Epoch 4/5
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3146 - accuracy: 0.8850
Epoch 5/5
313/313 [==============================] - 0s 1ms/step - loss: 0.3474 - accuracy: 0.8736


#**Result :**
89% Accuracy on training data . <br>
87% on test data.


#**Training with convolution**
Input image is $28*28$ pixel.<br>

###**Model structure:** <br>
64, number of $3*3$ filter .<br>
$2*2$ max pooling.<br>
64, $3*3$ filter.<br>
$2*2$ pax pooling.

128 fully connected neural network.<br>
10 classification in output fuction.

##**Note :**
Step 1 is to gather the data. You'll notice that there's a bit of a change here in that the training **data needed to be reshaped**. That's because the **first convolution expects** a **single tensor** containing everything, so instead of 60,000 28x28x1 items in a list, we have ***a single 4D list*** that is 60,000x28x28x1, and the same for the test images. If you don't do this, you'll get an error when training as the Convolutions do not recognize the shape.

In [9]:
import tensorflow as tf

mnist = tf.keras.datasets.fashion_mnist

(training_images,training_labels),(test_images,test_labels)= mnist.load_data()

#reshape training image from 3 channel(color image) to 1 channel(grey scale)
#there are 60000 image in training set, each 28*28 pizel
training_images = training_images.reshape(60000, 28,28,1)
training_images = training_images/255.0

#reshape test image
test_images = test_images.reshape(10000, 28,28,1)
test_images = test_images/255.0
model = tf.keras.models.Sequential([
  tf.keras.layers.Conv2D(64,(3,3),activation='relu',input_shape=(28,28,1)),
  tf.keras.layers.MaxPool2D(2,2),
  tf.keras.layers.Conv2D(64,(3,3),activation='relu'),
  tf.keras.layers.MaxPool2D(2,2),
  tf.keras.layers.Flatten(),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(10, activation ='softmax') 
])

model.compile(optimizer ='adam', loss='sparse_categorical_crossentropy',metrics=['accuracy'])
model.summary()


Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_7 (Conv2D)            (None, 26, 26, 64)        640       
_________________________________________________________________
max_pooling2d_6 (MaxPooling2 (None, 13, 13, 64)        0         
_________________________________________________________________
conv2d_8 (Conv2D)            (None, 11, 11, 64)        36928     
_________________________________________________________________
max_pooling2d_7 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
flatten_2 (Flatten)          (None, 1600)              0         
_________________________________________________________________
dense_4 (Dense)              (None, 128)               204928    
_________________________________________________________________
dense_5 (Dense)              (None, 10)               

In [10]:
#Training
model.fit(training_images,training_labels,epochs= 5)

#Test
test_loss = model.evaluate(test_images,test_labels)

Epoch 1/5
1875/1875 [==============================] - 76s 40ms/step - loss: 0.5962 - accuracy: 0.7814
Epoch 2/5
1875/1875 [==============================] - 75s 40ms/step - loss: 0.3005 - accuracy: 0.8896
Epoch 3/5
1875/1875 [==============================] - 75s 40ms/step - loss: 0.2413 - accuracy: 0.9108
Epoch 4/5
1875/1875 [==============================] - 75s 40ms/step - loss: 0.2124 - accuracy: 0.9214
Epoch 5/5
313/313 [==============================] - 4s 11ms/step - loss: 0.2570 - accuracy: 0.9069


#**Result using convolution :**
93% accuracy on training data. <br>
90% accuracy on test data.

# Visualizing the Convolutions and Pooling

This code will show us the convolutions graphically. The print (test_labels[;100]) shows us the first 100 labels in the test set, and you can see that the ones at index 0, index 23 and index 28 are all the same value (9). They're all shoes. Let's take a look at the result of running the convolution on each, and you'll begin to see common features between them emerge. Now, when the DNN is training on that data, it's working with a lot less, and it's perhaps finding a commonality between shoes based on this convolution/pooling combination.

In [13]:
print(test_labels[:100])

[9 2 1 1 6 1 4 6 5 7 4 5 7 3 4 1 2 4 8 0 2 5 7 9 1 4 6 0 9 3 8 8 3 3 8 0 7
 5 7 9 6 1 3 7 6 7 2 1 2 2 4 4 5 8 2 2 8 4 8 0 7 7 8 5 1 1 2 3 9 8 7 0 2 6
 2 3 1 2 8 4 1 8 5 9 5 0 3 2 0 6 5 3 6 7 1 8 0 1 4 2]


In [ ]:
import matplotlib.pyplot as plt

f,axarr = plt.subplot(3,4)

**EXERCISES**

1. Try editing the convolutions. Change the **32s** to either** 16 or 64**. What impact will this have on accuracy and/or training time.

2. **Remove the final Convoluti**on. What impact will this have on accuracy or training time?

3. **How about adding more Convolutions**? What impact do you think this will have? Experiment with it.

4. Remove all Convolutions but the first. What impact do you think this will have? Experiment with it. 

5. In the previous lesson you **implemented a callback** to check on the loss function and to cancel training once it hit a certain amount. See if you can implement that here!

1. **Using 32 filter.**

In [16]:
import tensorflow as tf

mnist = tf.keras.datasets.fashion_mnist

(training_images,training_labels),(test_images,test_labels) = mnist.load_data()

training_images = training_images.reshape(60000,28,28,1)
training_images = training_images/255.0

test_images = test_images.reshape(10000,28,28,1)
test_images = test_images/255.0

model = tf.keras.models.Sequential([
    tf.keras.layers.Conv2D(32, (3,3),activation='relu', input_shape=(28,28,1)),
    tf.keras.layers.MaxPool2D(pool_size=(2,2)),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(128,activation='relu'),
    tf.keras.layers.Dense(10,activation='softmax')
])

#model optimize
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])

#model train
model.fit(training_images,training_labels,epochs=10)

test_loss, test_acc = model.evaluate(test_images,test_labels)

#print( 'accuracy is : ' +test_acc)


Epoch 1/10
1875/1875 [==============================] - 33s 17ms/step - loss: 0.5263 - accuracy: 0.8134
Epoch 2/10
1875/1875 [==============================] - 32s 17ms/step - loss: 0.2668 - accuracy: 0.9039
Epoch 3/10
1875/1875 [==============================] - 32s 17ms/step - loss: 0.2176 - accuracy: 0.9213
Epoch 4/10
1875/1875 [==============================] - 32s 17ms/step - loss: 0.1841 - accuracy: 0.9320
Epoch 5/10
1875/1875 [==============================] - 32s 17ms/step - loss: 0.1548 - accuracy: 0.9433
Epoch 6/10
1875/1875 [==============================] - 32s 17ms/step - loss: 0.1361 - accuracy: 0.9491
Epoch 7/10
1875/1875 [==============================] - 32s 17ms/step - loss: 0.1154 - accuracy: 0.9582
Epoch 8/10
1875/1875 [==============================] - 32s 17ms/step - loss: 0.0964 - accuracy: 0.9640
Epoch 9/10
1875/1875 [==============================] - 32s 17ms/step - loss: 0.0803 - accuracy: 0.9708
Epoch 10/10
313/313 [==============================] - 2s 6ms/st

TypeError: ignored

In [18]:
print(test_acc)

0.9190999865531921


#Verdict :

97% accuracy in training data. <br>

91% acccuracy in test data.